### PR TITLE
feat: add @eslint/v9-to-v10-config codemod

### PR DESCRIPTION
**Pull request by khadatkarrohit**
## Summary

ESLint v10 removes several configuration mechanisms that were deprecated or made obsolete by the mandatory flat config format. Projects that still reference `/* eslint-env */` comments, `ESLINT_USE_FLAT_CONFIG`, `v10_config_lookup_from_file`, or the removed CLI flags (`--no-eslintrc`, `--env`, `--rulesdir`, `--ignore-path`, `--resolve-plugins-relative-to`) will break after upgrading.

This PR adds a new codemod `@eslint/v9-to-v10-config` that automatically removes all of those so any project can upgrade without manual search-and-replace across scripts, CI configs, and source files.

## What it transforms

**`eslint-env` inline comments — now a lint error in v10**

```js
// Before
/* eslint-env browser */
const el = document.getElementById('app')

/* eslint-env node, browser */
function read() { return process.env.NODE_ENV }

// After
const el = document.getElementById('app')

function read() { return process.env.NODE_ENV }
```

**Legacy env vars and CLI flags**

| Before | After |
| --- | --- |
| `ESLINT_USE_FLAT_CONFIG=true eslint .` | `eslint .` |
| `ESLINT_USE_FLAT_CONFIG=false eslint .` | `eslint .` |
| `export ESLINT_USE_FLAT_CONFIG=true && eslint .` | `eslint .` |
| `cross-env ESLINT_USE_FLAT_CONFIG=true eslint .` | `cross-env eslint .` |
| `v10_config_lookup_from_file` in `ESLINT_FLAGS` | Removed |
| `unstable_config_lookup_from_file` in `ESLINT_FLAGS` | Removed |
| `--no-eslintrc` | Removed |
| `--env browser` | Removed |
| `--rulesdir ./rules` | Removed |
| `--ignore-path .gitignore` | Removed |
| `--resolve-plugins-relative-to .` | Removed |

## How to use

```bash
npx codemod @eslint/v9-to-v10-config
```

## Testing

6 fixtures covering all transform cases. Validated against 30 real-world scenarios including edge cases found in production repos (microsoft/pyright, strvcom/code-quality-tools, gradio-app/gradio).

Full scenario results: https://github.com/khadatkarrohit/eslint-v9-ruletester-test/blob/main/SCENARIOS-config.md

```bash
pnpm --filter @eslint/v9-to-v10-config test
pnpm --filter @eslint/v9-to-v10-config check-types
```

All tests pass.

## Related

- Part of the v9→v10 series: companion to #8 (`@eslint/v9-to-v10-custom-rules`) and #9 (`@eslint/v9-to-v10-ruletester`)

### CODE REVIEW

File: codemods/v10/config/tests/eslint-env-basic/input.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/* eslint-env browser */
+const el = document.getElementById('app')
+
+/* eslint-env node, browser */
```
Comment (DMartens) [inline codemods/v10/config/tests/eslint-env-basic/input.js]: I think this fix is incomplete as this will likely lead to reports from rules dependening on the globals like `no-unused-vars`.
Either do not remove them and let the user fix them as `eslint-env` will be reported or also add the globals to the `eslint.config.js`.

Comment (khadatkarrohit) [reply codemods/v10/config/tests/eslint-env-basic/input.js]: You’re right. We’ve removed the eslint-env comment removal step entirely. ESLint v10 will report these comments, and users can migrate them manually.

File: codemods/v10/config/workflow.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        js-ast-grep:
+          language: javascript
+          js_file: scripts/remove-legacy-flags.ts
+          include:
+            - '**/*.ts'
+            - '**/*.tsx'
+            - '**/*.js'
+            - '**/*.jsx'
+            - '**/*.cjs'
+            - '**/*.mjs'
```
Comment (lumirlumir) [inline codemods/v10/config/workflow.yaml]:
```suggestion
            - '**/*.js'
            - '**/*.mjs'
            - '**/*.cjs'
            - '**/*.jsx' # Unlikely to be used for custom rules, but just in case
            - '**/*.ts'
            - '**/*.mts'
            - '**/*.cts'
            - '**/*.tsx' # Unlikely to be used for custom rules, but just in case
          exclude:
            - '**/*.d.ts'
            - '**/*.d.mts'
            - '**/*.d.cts'
```

https://github.com/eslint/codemods/pull/9#discussion_r3366905927

I think it would be nice if we could unify the extensions as mentioned in the review above.

**Review by lumirlumir — CHANGES_REQUESTED** [review]

**Comment by lumirlumir** [comment]
I think the current transform implementation also follows direct source code editing instead of using `SgNode#commitEdits(edits)`, as mentioned in https://github.com/eslint/codemods/pull/9#discussion_r3367362660.

It would be nice if we could consistently use the `SgNode#commitEdits(edits)` pattern here too, to prevent index-shift issues: once an earlier range is removed, later source indexes no longer point to the same text.

**Comment by khadatkarrohit** [comment]
> #9 (comment)

Thanks for flagging this! This transform is actually regex-based (replaceAll) rather than position-based, so it doesn't have the index-shift problem. SgRoot is only used to get the raw source text, no findAll or SgNode ranges involved, so commitEdits doesn't apply here. anything i am missing here?

File: codemods/v10/config/workflow.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+          js_file: scripts/remove-legacy-flags.ts
+          include:
+            - '**/*.json'
+      - id: remove-legacy-flags-yaml
+        name: 'Remove legacy env vars and CLI flags (YAML)'
+        depends_on:
+          - remove-legacy-flags-json
+        js-ast-grep:
+          language: yaml
+          js_file: scripts/remove-legacy-flags.ts
```
Comment (AugustinMauroy) [inline codemods/v10/config/workflow.yaml]: should be a different script that use yaml ast

File: codemods/v10/config/workflow.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            - '**/*.d.ts'
+            - '**/*.d.mts'
+            - '**/*.d.cts'
+      - id: remove-legacy-flags-json
+        name: 'Remove legacy env vars and CLI flags (JSON)'
+        depends_on:
+          - remove-legacy-flags-js
+        js-ast-grep:
+          language: json
+          js_file: scripts/remove-legacy-flags.ts
```
Comment (AugustinMauroy) [inline codemods/v10/config/workflow.yaml]: ditto

File: codemods/v10/config/scripts/remove-legacy-flags.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  for (const flag of FLAGS_WITH_VALUE) {
+    const escaped = flag.replaceAll(/[.*+?^${}()|[\]\\]/g, '\\$&')
+    // Remove the flag and its value; value stops at quotes to avoid consuming string delimiters
+    result = result.replaceAll(new RegExp(`\\s*${escaped}(\\s+${VALUE_TOKEN})?`, 'g'), '')
+  }
+
+  return result
+}
+
+export default async function transform(root: SgRoot): Promise<string | null> {
```
Comment (AugustinMauroy) [inline codemods/v10/config/scripts/remove-legacy-flags.ts]:
```suggestion
export default async function transform(root: SgRoot<JS>): Promise<string | null> {
```

**Review by AugustinMauroy — CHANGES_REQUESTED** [review]
you must not do string transformation

**Comment by lumirlumir** [comment]
I see that this uses `commitEdits`, but the transform still looks too broad. It uses ast-grep only to find string/scalar nodes, then applies regex replacements inside every string.

That means we aren’t matching semantic locations like the `scripts` field in `package.json`, `run` steps in GitHub Actions, or command strings in JS `execSync()` calls. This can rewrite unrelated documentation strings, snapshots, config values, or YAML block text.

I think we should either narrow the AST matches to known command-bearing fields or call sites, or use structured selectors for the relevant formats instead of running global string replacements over arbitrary literals.

Also, the current regex pattern approach doesn’t seem like a good fit, as it would make future maintenance more difficult.

**Comment by khadatkarrohit** [comment]
> I see that this uses `commitEdits`, but the transform still looks too broad. It uses ast-grep only to find string/scalar nodes, then applies regex replacements inside every string.
> 
> That means we aren’t matching semantic locations like the `scripts` field in `package.json`, `run` steps in GitHub Actions, or command strings in JS `execSync()` calls. This can rewrite unrelated documentation strings, snapshots, config values, or YAML block text.
> 
> I think we should either narrow the AST matches to known command-bearing fields or call sites, or use structured selectors for the relevant formats instead of running global string replacements over arbitrary literals.
> 
> Also, the current regex pattern approach doesn’t seem like a good fit, as it would make future maintenance more difficult.

The current transforms operate on string/scalar nodes without semantic filtering, which can affect snapshots and documentation strings.

I’ll narrow the YAML transform to run: step values and the JSON transform to scripts fields using structured selectors. JS is trickier since commands can appear in many APIs (execSync, spawnSync, execa, etc.), but I’ll at least exclude obvious non-command contexts such as snapshots and test fixtures and iterate from there.

**Comment by AugustinMauroy** [comment]
> JS is trickier since commands can appear in many APIs (execSync, spawnSync, execa, etc.)

IMO you (eslint) should only catch node api, other thing can be "skipped"

**Comment by khadatkarrohit** [comment]
Force push was needed because the branch was rebased onto origin/main.

File: codemods/v10/config/scripts/legacy-flag-transforms.ts
Comment (AugustinMauroy) [inline codemods/v10/config/scripts/legacy-flag-transforms.ts:1]: this file can be removed ?

Comment (khadatkarrohit) [reply codemods/v10/config/scripts/legacy-flag-transforms.ts:1]: No, this file is the shared transform library used by remove-legacy-flags.ts, remove-legacy-flags-json.ts, and remove-legacy-flags-yaml.ts. It contains the core logic that all three transforms rely on.

Was there something else you were referring to?

Comment (AugustinMauroy) [reply codemods/v10/config/scripts/legacy-flag-transforms.ts:1]: My bad that's right

File: .lintstagedrc.mjs [hunk 6]
```diff
@@ -11,7 +11,8 @@ export default {
   '*.{ts,mts,cts,tsx}': 'oxlint --type-aware --type-check --fix',
   /** @param {string[]} files */
   '*.{json,yaml,yml}': (files) => {
-    const filtered = excludeLockfiles(files)
+    // oxfmt errors when all files are excluded by ignorePatterns; skip test fixtures upfront
+    const filtered = excludeLockfiles(files).filter((f) => !/[/\\]tests[/\\]/.test(f))
     return filtered.length ? [`oxfmt --write ${filtered.join(' ')}`] : []
```
Comment (lumirlumir) [inline .lintstagedrc.mjs]:
```suggestion
    return filtered.length ? [`oxfmt --write --no-error-on-unmatched-pattern ${filtered.join(' ')}`] : []
```

What I found when I worked on https://github.com/eslint/codemods/pull/25 is that we can simply add the `--no-error-on-unmatched-pattern` flag to avoid the problem you encountered.

**Review by lumirlumir — CHANGES_REQUESTED** [review]

**Review by alexbit-codemod — APPROVED** [review]
